atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/CacheMetrics.java

```diff
@@ -0,0 +1,96 @@
+
+import com.palantir.atlasdb.AtlasDbMetricNames;
+import com.palantir.atlasdb.util.AccumulatingValueMetric;
+import com.palantir.atlasdb.util.CurrentValueMetric;
+import com.palantir.atlasdb.util.MetricsManager;
+
+public final class CacheMetrics {
+    private final AccumulatingValueMetric hits;
+    private final AccumulatingValueMetric misses;
+    private final AccumulatingValueMetric cacheSize;
+    private final CurrentValueMetric<Integer> eventCacheValidationFailures;
+    private final CurrentValueMetric<Integer> valueCacheValidationFailures;
+    private final MetricsManager metricsManager;
+
+    private CacheMetrics(
+            AccumulatingValueMetric hits,
+            AccumulatingValueMetric misses,
+            AccumulatingValueMetric cacheSize,
+            CurrentValueMetric<Integer> eventCacheValidationFailures,
+            CurrentValueMetric<Integer> valueCacheValidationFailures,
+            MetricsManager metricsManager) {
+        this.hits = hits;
+        this.misses = misses;
+        this.cacheSize = cacheSize;
+        this.eventCacheValidationFailures = eventCacheValidationFailures;
+        this.valueCacheValidationFailures = valueCacheValidationFailures;
+        this.metricsManager = metricsManager;
+    }
+
+    public static CacheMetrics create(MetricsManager metricsManager) {
+        return new CacheMetrics(
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class, AtlasDbMetricNames.CACHE_HITS, AccumulatingValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class, AtlasDbMetricNames.CACHE_MISSES, AccumulatingValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class, AtlasDbMetricNames.CACHE_SIZE, AccumulatingValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class,
+                        AtlasDbMetricNames.EVENT_CACHE_VALIDATION_FAILURES,
+                        CurrentValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class,
+                        AtlasDbMetricNames.VALUE_CACHE_VALIDATION_FAILURES,
+                        CurrentValueMetric::new),
+                metricsManager);
+    }
+
+    public void registerHits(long number) {
+        hits.accumulateValue(number);
+    }
+
+    public void registerMisses(long number) {
+        misses.accumulateValue(number);
+    }
+
+    public void increaseCacheSize(long added) {
+        cacheSize.accumulateValue(added);
+    }
+
+    public void decreaseCacheSize(long removed) {
+        cacheSize.accumulateValue(-removed);
+    }
+
+    public void registerEventCacheValidationFailure() {
+        eventCacheValidationFailures.setValue(1);
+    }
+
+    public void registerValueCacheValidationFailure() {
+        valueCacheValidationFailures.setValue(1);
```

Comment:
Interesting. For these two, why did we choose to use a null/1 value as opposed to an accumulating value metric?

Might want to rename this to `hasFailures` or something (to me at least the intuitive reading of a 1 value is "there was one failure", as opposed to "we're in a failed state").

Reply:
(I see this is a straight port of the previous system. Feel free to skip for this PR.)

Reply:
It's because as soon as you have a failure you fall back to no cache, so you will never have more